kong/router/atc_compat.lua

```diff
@@ -547,13 +564,62 @@ function _M:exec(ctx)
 
   req_uri = normalize(req_uri, true)
 
-  local match_t = self:select(req_method, req_uri, req_host, req_scheme,
-                              nil, nil, nil, nil,
-                              sni, headers)
+  local headers_key do
+    local headers_count = 0
+    for name, value in pairs(headers) do
+      local name = name:gsub("-", "_"):lower()
+
+      if type(value) == "table" then
+        for i = 1, #value do
+          value[i] = value[i]:lower()
+        end
+        tb_sort(value)
+        value = tb_concat(value, ", ")
+
+      else
+        value = value:lower()
+      end
+
+      if headers_count == 0 then
+        headers_key = { "|", name, "=", value }
+
+      else
+        headers_key[headers_count+1] = "|"
+        headers_key[headers_count+2] = name
+        headers_key[headers_count+3] = "="
+        headers_key[headers_count+4] = value
+      end
+
+      headers_count = headers_count + 4
+    end
+
+    headers_key = headers_key and tb_concat(headers_key, nil, 1, headers_count) or ""
+  end
+
+  -- cache lookup
+
+  local cache_key = (req_method or "") .. "|" .. (req_uri or "") ..
+                    "|" .. (req_host or "") .. "|" .. (sni or "") ..
+                    headers_key
+  local match_t = self.cache:get(cache_key)
   if not match_t then
-    return
+    if self.cache_neg:get(cache_key) then
+      return nil
```

Comment:
I think `return` is enough, because below code also did so.

Reply:
I want to be a bit more explicit, we are actually returning only `nil` to indicate a not match condition. It is better style IMO.